modules/katamari/src/main/ts/ephox/katamari/api/Result.ts

```diff
@@ -1,156 +1,288 @@
+   * anything in the case where the `Result` doesn't have a value inside it. If
+   * you're not sure whether your use-case fits into transforming **into**
+   * something or **doing** something, check whether it has a return value. If
+   * it does, you should be performing a transform.
+   */
+  readonly each: (worker: (value: T) => void) => void;
+
+  /**
+   * Convert this `Result<T, E>` to an `Optional<T>`. If this object contains a
+   * value, so will the output. If this object contains an error, the error will
+   * be discarded and the output will not contain a value.
+   */
+  readonly toOptional: () => Optional<T>;
+}
+
+/**
+ * Debugging information. This is deliberately not included in the exported
+ * types, because we don't want production code to use it, but it's very useful
+ * to have this information included in the objects themselves for when you're
+ * looking at them in the console or debugger.
+ */
+interface DebugResult<T, E> extends Result<T, E> {
+  /**
+   * True for `Value`, False for `Error`.
+   */
+  readonly tag: boolean;
 
-  const fold = <U>(onError: (err: E) => U, _: (value: T) => U) => {
-    return onError(message);
+  /**
+   * The value or error contained within this Result object.
+   */
+  readonly inner: T | E;
+}
+
+/**
+ * Creates a new `Result<T, E>` that **does** contain a value.
+ */
+const value = <T, E = never>(value: T): Result<T, E> => {
+
+  const applyHelper = <U>(fn: (value: T) => U): U => fn(value);
+  const constHelper = Fun.constant(value);
+  const outputHelper = <E2 = E>() => output as unknown as Result<T, E2>;
+
+  const output: DebugResult<T, E> = {
+    // Debug info
+    tag: true,
+    inner: value,
+    // Actual Result methods
+    fold: (_onError, onValue) => onValue(value),
+    isValue: Fun.always,
+    isError: Fun.never,
+    map: (mapper) => Result.value(mapper(value)),
+    mapError: outputHelper,
+    bind: applyHelper,
+    exists: applyHelper,
+    forall: applyHelper,
+    getOr: constHelper,
+    or: outputHelper,
+    getOrThunk: constHelper,
+    orThunk: outputHelper,
+    getOrDie: constHelper,
+    each: (fn) => {
+      // Can't write the function inline because we don't want to return something by mistake
+      fn(value);
+    },
+    toOptional: () => Optional.some(value),
   };
 
-  return {
+  return output;
+};
+
+/**
+ * Creates a new `Result<T, E>` that **does not** contain a value, and therefore
+ * contains an error.
+ */
+const error = <T = never, E = any>(error: E): Result<T, E> => {
+
+  const output: DebugResult<T, E> = {
+    // Debug info
+    tag: false,
+    inner: error,
+    // Actual Result methods
+    fold: (onError, _onValue) => onError(error),
     isValue: Fun.never,
     isError: Fun.always,
-    getOr: Fun.identity,
-    getOrThunk,
-    getOrDie,
-    or,
-    orThunk,
-    fold,
-    map,
-    mapError,
-    each: Fun.noop,
-    bind,
+    map: <U>(_mapper) => output as unknown as Result<U, E>,
```

Comment:
This and `bind` looks like it could also use the `outputHelper` function from `value`?